### ISSUE TITLE
loadMetaContainers poor performance (70ms+) on every request (including control panel)

### ISSUE THREAD

**Issue by jasonmccallister** [nystudio107/craft-seomatic#954]
### Describe the bug
When loading the control panel, `loadMetaContainers`, is the slowest part of the page and seems unnecessary. 

https://github.com/nystudio107/craft-seomatic/blob/develop/src/services/MetaContainers.php#L157-L243

### To reproduce
Steps to reproduce the behaviour:
1. Go to any control panel page
2. Load it
3. Look at the debug > profiling bar and see `loadMetaContainers`

### Expected behaviour
Not sure of the intent of this method, but since its slow perhaps it should only execute when required? 

### Screenshots
<img width="1223" alt="CleanShot 2021-08-20 at 09 57 22@2x" src="https://user-images.githubusercontent.com/5354908/130244519-2aae70d1-8576-477d-9f06-e8814b71287a.png">
<img width="1134" alt="CleanShot 2021-08-20 at 09 57 00@2x" src="https://user-images.githubusercontent.com/5354908/130244528-2edd1128-1527-4538-955d-71aec163aaa3.png">

### Versions
- Plugin version: 3.4.1
- Craft version: 3.7.9

**Comment by khalwat** [comment]
There's a two-level cache in place, so I just want to confirm that you've profiled the same page(s) multiple times? With `devMode` on, SEOmatic's cache is only 30 seconds as well (in production, it's infinite).

This call is needed on any page that could have a SEOmatic preview on it (so Entries, Products, Category Groups, and so on). If you can provide a method that allows me to determine whether we're on such a CP page, I could look into the ramifications of utilizing it.

**Comment by jasonmccallister** [comment]
`devMode` is off and reloaded the dashboard twice, both times ~80ms.

Tried the following in the control panel:

- dashboard (80ms)
- utilities/system-report (76ms)
- utilities/clear-cache (84ms)
- utilities/deprecation-errors (46ms)
- utilities/system-messages (77ms)

**Comment by khalwat** [comment]
Cool, sounds like a good sample then... if you can provide a method that allows me to determine whether we're on a CP page when entries, products, etc. can be displayed, I could look into the ramifications of utilizing it to shave off this ~80ms

**Comment by jasonmccallister** [comment]
>  if you can provide a method that allows me to determine whether we're on a CP page when entries, products, etc. can be displayed

Not quite following your statement, do you mean "you" as in P&T provide a method in the Craft core codebase?

**Comment by khalwat** [comment]
Right, if you're aware of such a method existing, let me know. I'm not aware of it, so I'd be left with the case where in some places in the CP, the meta containers actually do need to be loaded already, and the way it's wired up, it's doing it early in the process (in the Twig extension) because they need to be in place before any Twig templates are rendered.

Many parts of the system depend on these meta containers being loaded & instantiated already.

It's fairly integrated in how the plugin works, so any good way of detecting if we're on a page in the CP with sidebar previews would be a huge help in addressing this.

**Comment by khalwat** [comment]
So just as a test, I've pushed a version that excludes the method from loading if it's a CP request: https://github.com/nystudio107/craft-seomatic/commit/a6ee640127feb695847bd94e4eb7c8a3f50c1cd7

I still need to do a bunch of testing on it (there are a number of places where this could go wrong, as mentioned above), but it looks like I'm wrong, and maybe we can just remove the explicit call completely for CP requests.

This is because it appears that at least in the case of the previews, it is calling `loadMetaContainers()` itself on those pages. My concern is other places in the meta rendering system that may depend on the containers being loaded already, but so far so good!

You can try it now by setting your semver in your `composer.json` to look like this:
```json
    "nystudio107/craft-seomatic": "dev-develop as 3.4.3”,
```
Then do a `composer clear-cache && composer update`

**Comment by jasonmccallister** [comment]
I was going to recommend `getIsCpRequest` but looks like you added it :)

**Comment by khalwat** [comment]
Right, the problem is there _are_ places in the CP where the meta containers need to be loaded. I'll be going through and testing to make sure there are no ripple-effect regressions from this change.

**Comment by angrybrad** [comment]
If it needs to load in certain places in the control panel, probably check if it's a CP request along with a list of whitelisted CP routes.

**Comment by khalwat** [comment]
Well, the whitelisted routes are a little difficult, in that it needs to be available on any element editing page, so entries, products, and third party elements like calendar events.

**Comment by angrybrad** [comment]
Ahhh... maybe grab a list of all registered elements and dynamically create a whitelist?

**Comment by khalwat** [comment]
Right, that's what I was saying earlier @angrybrad in that I'm not aware of any good way to actually do that.

But it may be a moot point, it may be possible to just not call `loadMetaContainers()` for CP requests, I just need to do more testing to ensure no regressions from the change.

**Comment by khalwat** [comment]
Addressed in: https://github.com/nystudio107/craft-seomatic/commit/a6ee640127feb695847bd94e4eb7c8a3f50c1cd7

You can try it now by setting your semver in your `composer.json` to look like this:
```json
    "nystudio107/craft-seomatic": "dev-develop as 3.4.3”,
```
Then do a `composer clear-cache && composer update`

**Comment by khalwat** [comment]
Unfortunately, it may end up being that I need to revert this change. People are using email templates in the CP where they expect the various metacontainers to be loaded, and I'm not seeing any events I can hook into to lazily load the metacontainers just before an email would be rendered, for instance.

**Comment by OwenMelbz** [comment]
Not sure which issue is best to respond to with this but we've got another issue semi-related to this:

I'm *guessing* because the loadMetaContainers doesn't load in the CP now, when you click "Preview" or "View" from editing an entry, you get an internal server error because some things have not been initialised.

e.g. We try to use the `seomatic.jsonLd.get('breadcrumbList').itemListElement` to loop around the crumbs.

I'm guessing as these are now not loaded in the CP causing the issue.

When you use the "View" button that generates a `?x-craft-preview=XXXXX` link it errors, when you delete the query string (thus making it a non-cp request) it works fine.

![yii_base_ErrorException](https://user-images.githubusercontent.com/1094740/132321879-2e2d088f-9ccd-4dfc-a101-540644cb7d57.png)

**Comment by khalwat** [comment]
This is in your custom code yes? I think you probably should be checking for `null` here, e.g.: 

```twig
{% set crumbs = seomatic.jsonLd.get('breadcrumbList') %}
{% if crumbs | length %}
    {# Do your stuff #}
{% endif %}
```

Or in PHP: 
```php
$crumbs = Seomatic::$plugin->jsonLd->get('breadcrumbList');
if ($crumbs) {
// Do you stuff
}
```

I realize this change comes from the containers not being loaded yet in the CP, but there are other cases where the various meta items could be `null`. I will also look at the code to see if there's a place where the containers could be re-injected

**Comment by khalwat** [comment]
Addressed in: https://github.com/nystudio107/craft-seomatic/commit/b1a4c9383a9c99125d299a79a78fdac6ed9275a0

You can try it now by setting your semver in your `composer.json` to look like this:
```json
    "nystudio107/craft-seomatic": "dev-develop as 3.4.7”,
```
Then do a `composer clear-cache && composer update`
